lib/network/level-net/sio-stream.js

```diff
@@ -0,0 +1,372 @@
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._onerror(error);
+            delete this.remoteStreams[streamId];
+        });
+
+        this.socket.on('stream-hangup', payload => {
+            const { streamId } = payload;
+            debug('received stream-hangup event streamid', streamId);
+            const stream = this.localStreams[streamId];
+            if (!stream) {
+                debug('no local stream registered with id', streamId,
+                      '(may have already reached the end)');
+                return;
+            }
+            this.destroyStream(streamId);
+        });
+    }
+
+    encodeStreams(arg) {
+        if (!arg) {
+            return arg;
+        }
+        const isReadStream = (typeof(arg.pipe) === 'function'
+                              && typeof (arg.read) === 'function');
+        const isWriteStream = (typeof(arg.write) === 'function');
+
+        if (isReadStream || isWriteStream) {
+            const streamId = uuid();
+            const encodedStream = {
+                $streamId: streamId,
+                readable: isReadStream,
+                writable: isWriteStream,
+            };
+            debug('encoded stream:', encodedStream);
+            let transportStream;
+            if (isReadStream && isWriteStream) {
+                transportStream = new SIODuplexStream(this, streamId,
+                                                      this.maxPendingAck,
+                                                      this.ackTimeoutMs);
+            } else if (isReadStream) {
+                transportStream = new SIOOutputStream(this, streamId,
+                                                      this.maxPendingAck,
+                                                      this.ackTimeoutMs);
+            } else {
+                transportStream = new SIOInputStream(this, streamId);
+            }
+            if (isReadStream) {
+                arg.pipe(transportStream);
+                this.localStreams[streamId] = arg;
+                arg.on('end', () => {
+                    debug('stream end, removing from local streams',
+                          streamId);
+                    delete this.localStreams[streamId];
+                });
+            }
+            if (isWriteStream) {
+                transportStream.pipe(arg);
+            }
+            return encodedStream;
+        }
+        if (typeof(arg) === 'object') {
+            let encodedObj;
+            if (Array.isArray(arg)) {
+                encodedObj = [];
+            } else {
+                encodedObj = {};
+            }
+            // user objects are simple flat objects and we want to
+            // copy all their properties
+            // eslint-disable-next-line
+            for (const k in arg) {
+                encodedObj[k] = this.encodeStreams(arg[k]);
```

Comment:
It will preserve the order, because the for loop will iterate over array "keys" 0, 1, ...